lib/gruf/interceptors/instrumentation/request_logging/interceptor.rb

```diff
@@ -0,0 +1,198 @@
+# coding: utf-8
+# Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+require 'socket'
+require_relative 'formatters/base'
+require_relative 'formatters/logstash'
+require_relative 'formatters/plain'
+
+module Gruf
+  module Interceptors
+    module Instrumentation
+      module RequestLogging
+        ##
+        # Represents an error if the formatter does not extend the base formatter
+        #
+        class InvalidFormatterError < StandardError; end
+
+        ##
+        # Handles Rails-style request logging for gruf services.
+        #
+        # This is added by default to gruf servers; if you have `Gruf.use_default_hooks = false`, you can add it back
+        # manually by doing:
+        #
+        #   Gruf::Instrumentation::Registry.add(:request_logging, Gruf::Instrumentation::RequestLogging::Hook)
+        #
+        class Interceptor < ::Gruf::Interceptors::ServerInterceptor
+
+          ###
+          # Log the request, sending it to the appropriate formatter
+          #
+          # @return [String]
+          #
+          def call
+            result = Gruf::Interceptors::Timer.time do
+              yield
+            end
+
+            if result.successful?
+              type = options.fetch(:success_log_level, :info).to_sym
+              status_name = 'GRPC::Ok'
+            else
+              type = options.fetch(:failure_log_level, :error).to_sym
+              status_name = result.message_class_name
+            end
+
+            payload = {}
+            if !client_streamer?
```

Comment:
If you perform the suggested refactor to build a `RequestType` object, you can update this part of the code to as the `RequestType` directly if it is a `client_streamer?` instead of using the adhoc logic on L89.